pkg/broker/util.go

```diff
@@ -31,78 +31,37 @@ func getGlobalOverrides(brokerID string) map[string]string {
 }
 
 func prescribeOverrides(b AwsBroker, services []osb.Service) []osb.Service {
-	if !b.prescribeOverrides {
+	if !b.prescribeOverrides || len(b.globalOverrides) == 0 {
 		return services
 	}
-	// TODO: Alot of duplication of code with ServiceDefinitionToOsb, should cleanup
-	for s, service := range services {
-		for p, plan := range service.Plans {
-			overrideKeys := make([]string, 0)
-			for o := range b.globalOverrides {
-				overrideKeys = append(overrideKeys, o)
-			}
-			glog.Infoln(overrideKeys)
-			schemas := map[string]map[string]interface{}{
-				"create": plan.Schemas.ServiceInstance.Create.Parameters.(map[string]interface{}),
+
+	var overridenKeys []string
+	for k := range b.globalOverrides {
+		overridenKeys = append(overridenKeys, k)
+	}
+
+	for _, service := range services {
+		for _, plan := range service.Plans {
+			params := plan.Schemas.ServiceInstance.Create.Parameters.(map[string]interface{})
+			for _, k := range overridenKeys {
+				delete(params["properties"].(map[string]interface{}), k)
+				params["required"] = deleteFromSlice(params["required"].([]string), k)
 			}
+
 			if plan.Schemas.ServiceInstance.Update != nil {
-				schemas["update"] = plan.Schemas.ServiceInstance.Update.Parameters.(map[string]interface{})
-			}
-			for schemaName, schema := range schemas {
-				props := make(map[string]interface{})
-				required := make([]string, 0)
-				for k, v := range schema {
-					switch k {
-					case "properties":
-						for pk, pv := range v.(map[string]interface{}) {
-							if !stringInSlice(pk, overrideKeys) {
-								props[pk] = pv
-							}
-						}
-					case "required":
-						glog.Infoln(v)
-						for _, r := range v.([]string) {
-							if !stringInSlice(r, overrideKeys) {
-								required = append(required, r)
-							}
-						}
-					}
+				params := plan.Schemas.ServiceInstance.Update.Parameters.(map[string]interface{})
+				for _, k := range overridenKeys {
+					delete(params["properties"].(map[string]interface{}), k)
+					params["required"] = deleteFromSlice(params["required"].([]string), k)
 				}
-				if schemaName == "create" {
-					params := map[string]interface{}{
-						"type":       "object",
-						"properties": props,
-						"$schema":    "http://json-schema.org/draft-06/schema#",
-					}
-					if len(required) > 0 {
```

Comment:
cloudfoundry insists on required containing more than 0 slices, so we need to check whether this is empty after removing all the overrides and delete the key if it is.

Reply:
I'll add a test for this in the next couple of days

Reply:
@jaymccon does Cloud Foundry care if `properties` is an empty map? Or do we have to delete that too if all properties have been overridden?

Reply:
having an empty properties block is accepted as valid by CF.

tested with public S3 production plan, aws_access_key, aws_secret_key, region overridden. 

GET /v2/catalog returns:

```json
...
{
  "service_instance": {
    "create": {
      "parameters": {
        "$schema": "http://json-schema.org/draft-06/schema#",
        "properties": {},
        "type": "object"
      }
    }
  }
}
...
```

Reply:
Sweet. I added the bit for removing `required` if all required params have been overridden. I also added a test.